src/ssl.c

```diff
@@ -13539,6 +13539,53 @@ static int SslSessionCacheOff(const WOLFSSL* ssl, const WOLFSSL_SESSION* session
                 ;
 }
 
+#if defined(HAVE_SESSION_TICKET) && defined(WOLFSSL_TLS13) &&                  \
+    defined(WOLFSSL_TICKET_NONCE_MALLOC)
+/**
+ * SessionTicketNoncePrealloc() - prealloc a buffer for ticket nonces
+ * @output: [in] pointer to WOLFSSL_SESSION object that will soon be a
+ * destination of a session duplication
+ * @buf: [out] address of the preallocated buf
+ * @len: [out] len of the preallocated buf
+ *
+ * prealloc a buffer that will likely suffice to contain a ticket nonce. It's
+ * used when copying session under lock, when syscalls need to be avoided. If
+ * output already has a dynamic buffer, it's reused.
+ */
+static int SessionTicketNoncePrealloc(WOLFSSL_SESSION* output, byte** buf,
+    byte* len)
+{
+    byte* tmp;
+
+    *buf = NULL;
+    /* steal the buffer */
+    if (output->ticketNonce.data != output->ticketNonce.dataStatic) {
+        *buf = output->ticketNonce.data;
+        *len = output->ticketNonce.len;
+        output->ticketNonce.data = output->ticketNonce.dataStatic;
+        output->ticketNonce.len = 0;
+        if (output->ticketNonce.len >= PREALLOC_SESSION_TICKET_NONCE_LEN)
+            return 0;
+    }
+
+    tmp = XREALLOC(*buf, PREALLOC_SESSION_TICKET_NONCE_LEN, output->heap,
```

Comment:
I really would like us to avoid using XREALLOC. It is not portable and not available in all cases. However if we already require it for TLS v1.3 session tickets then it is okay. I suspect this creates a new XREALLOC requirement. Could we add an `#ifndef XREALLOC` fallback to XMALLOC, XMEMCPY and XFREE?

Reply:
removed one xrealloc and gated the other one